File: solutions/scala/alphametics/1/src/main/scala/Alphametics.scala
import MapUtilities.RichMap
import SetUtilities.RichHashSet

import scala.annotation.tailrec

object Alphametics {
  private val SingleDigitNumbers   : List[Int]                                     = (0 to 9)
    .toList
  private val calculateFactorWeight: List[Int] => PartialFunction[(Int, Int), Int] = permutation => {
    case (key, count) =>
      count * permutation(key)
  }

  def solve(equation: String): Option[Map[Char, Int]] = {
    val activeLetters     : Iterable[Char]  = extractUnknowns(equation)
    val activeLettersArray: Array[Char]     = activeLetters.toArray
    val activeLetterCount : Int             = activeLettersArray.length
    val mappedCharacters                    = IterableUtilities.mapCharactersToIntegers(activeLettersArray)
    val tokens            : List[List[Int]] = mappedCharacters
      .tokenize(equation)

    val columns : List[(Int, List[(Int, Int)])] = IterableUtilities.parse(tokens)
    val zeroMask: List[Boolean]                 = IterableUtilities.extractWordInitials(tokens)
                                                                   .buildZeroMask(activeLetterCount)

    ListUtilities.createCombinations(SingleDigitNumbers)(activeLetterCount)
                 .filter(canBeZero(zeroMask))
                 .find(columnSum(columns))
                 .map(toResultDictionary(activeLettersArray))
  }

  private def extractUnknowns(equation: String): Iterable[Char] =
    equation.filter(!" +=".contains(_))
            .distinct

  private def canBeZero(zeroMask: List[Boolean])(tokens: List[Int]): Boolean = {
    val found = tokens.indexWhere(_ == 0)
    found == -1 || zeroMask(found)
  }

  @tailrec
  private def columnSum(column: List[(Int, List[(Int, Int)])], carry: Int = 0)(permutation: List[Int]): Boolean =
    if (column.isEmpty) {
      carry == 0
    }
    else {
      val (y, xs) = column.head
      val sum     = calculatePermutationValue(carry, xs)(permutation)
      (permutation(y) == sum % 10) &&
      columnSum(column.tail, sum / 10)(permutation)
    }

  private def calculatePermutationValue(baseValue: Int, factorWeights: List[(Int, Int)])(permutation: List[Int]): Int =
    factorWeights.map(calculateFactorWeight(permutation)).sum + baseValue

  private def toResultDictionary(activeLetters: Array[Char])
                                (result: List[Int]): Map[Char, Int] =
    activeLetters.zip(result)
                 .toMap
}

File: solutions/csharp/alphametics/2/Alphametics.cs
using System;
using System.Collections.Generic;
using System.Linq;

// Derived from Martin Freedman's solution, but refactored to fit my coding preferences
// https://exercism.org/tracks/csharp/exercises/alphametics/solutions/martinfreedman

public class Alphametics
{
	private static readonly List<int> SingleDigitNumbers = Enumerable.Range(0, 10)
		.ToList();

	public static IDictionary<char, int> Solve(string equation)
	{
		IEnumerable<char> activeLetters = ExtractUnknowns(equation);
		int activeLetterCount = activeLetters.Count();
		List<List<int>> tokens = activeLetters.MapCharactersToIntegers()
			.Tokenise(equation);

		List<(int target, List<(int key, int count)>)> columns = tokens.Parse();
		List<bool> zeroMask = tokens.ExtractWordInitials()
			.BuildZeroMask(activeLetterCount);

		List<int> result = SingleDigitNumbers.CreateCombinations(activeLetterCount)
			.Where(tokenCharacters => CanBeZero(tokenCharacters, zeroMask))
			.Where(permutation => ColumnSum(columns, 0, permutation))
			.FirstOrDefault();

		return result != null
			? ToResultDictionary(activeLetters, result)
			: throw new ArgumentException(nameof(equation));
	}

	private static IEnumerable<char> ExtractUnknowns(string equation) =>
		equation.Where(c => !" +=".Contains(c))
			.Distinct();

	private static bool CanBeZero(List<int> tokens, List<bool> zeroMask)
	{
		int found = tokens.FindIndex(i => i == 0);
		return found == -1 || zeroMask[found];
	}

	private static bool ColumnSum(List<(int y, List<(int key, int count)> xs)> column, int carry, List<int> permutation)
	{
		List<(int y, List<(int key, int count)> xs)>.Enumerator enumerator = column.GetEnumerator();
		while (enumerator.MoveNext())
		{
			(int y, List<(int key, int count)> xs) = (enumerator.Current.y, enumerator.Current.xs);
			int sum = xs.Sum(k => k.count * permutation[k.key]) + carry;
			if (permutation[y] == sum % 10)
			{
				carry = sum / 10;
			}
			else
			{
				return false;
			}
		}

		return carry == 0;
	}

	private static Dictionary<char, int> ToResultDictionary(IEnumerable<char> activeLetters, List<int> result) =>
		result.Zip(
				activeLetters,
				(digit, letter) => (c: letter, i: digit)
		)
		.ToDictionary(
			keyValuePair => keyValuePair.c,
			keyValuePair => keyValuePair.i
		);
}